### ISSUE TITLE
AstGen crash when struct field missing type

### ISSUE THREAD

**Issue by andrewrk** [ziglang/zig#9206]
Caught by one of the compile-errors tests:

```zig
const Letter = struct {
    A,
};
export fn entry() void {
    var a = Letter { .A = {} };
}
```

Expected behavior:

```
$ ./zig ast-check test.zig
test.zig:2:5: error: struct field missing type
```

Actual behavior:

```
$ ./zig ast-check test.zig
thread 31896 panic: reached unreachable code
/home/andy/dev/zig/src/AstGen.zig:458:18: 0xf66cd2 in AstGen.expr (zig1)
        .root => unreachable, // Top-level declaration.
                 ^
/home/andy/dev/zig/src/AstGen.zig:251:16: 0xf66b9e in AstGen.typeExpr (zig1)
    return expr(gz, scope, .{ .ty = .type_type }, type_node);
               ^
/home/andy/dev/zig/src/AstGen.zig:3564:25: 0xe1ba60 in AstGen.structDeclInner (zig1)
            try typeExpr(&block_scope, &block_scope.base, member.ast.type_expr);
                        ^
/home/andy/dev/zig/src/AstGen.zig:3952:47: 0x10a76c3 in AstGen.containerDecl (zig1)
            const result = try structDeclInner(gz, scope, node, container_decl, layout);
                                              ^
/home/andy/dev/zig/src/AstGen.zig:818:33: 0xf6b6c1 in AstGen.expr (zig1)
            return containerDecl(gz, scope, rl, node, tree.containerDeclTwo(&buffer, node));
                                ^
/home/andy/dev/zig/src/AstGen.zig:3115:35: 0xf63efe in AstGen.globalVarDecl (zig1)
        const init_inst = try expr(
                                  ^
/home/andy/dev/zig/src/AstGen.zig:3515:37: 0xe1aff3 in AstGen.structDeclInner (zig1)
                astgen.globalVarDecl(gz, &namespace.base, &wip_decls, member_node, tree.simpleVarDecl(member_node)) catch |err| switch (err) {
                                    ^
/home/andy/dev/zig/src/AstGen.zig:107:31: 0xc6a188 in AstGen.generate (zig1)
    if (AstGen.structDeclInner(
                              ^
/home/andy/dev/zig/src/Module.zig:2483:35: 0xe89d17 in Module.astGenFile (zig1)
    file.zir = try AstGen.generate(gpa, file.tree);
                                  ^
/home/andy/dev/zig/src/Compilation.zig:2303:19: 0xe878e2 in Compilation.workerAstGenFile (zig1)
    mod.astGenFile(file) catch |err| switch (err) {
                  ^
/home/andy/dev/zig/src/ThreadPool.zig:117:13: 0xe8c898 in ThreadPool.Closure.runFn (zig1)
            @call(.{}, func, closure.arguments);
            ^
/home/andy/dev/zig/src/ThreadPool.zig:35:38: 0xc88477 in ThreadPool.Worker.run (zig1)
                (run_node.data.runFn)(&run_node.data);
                                     ^
/home/andy/dev/zig/lib/std/Thread.zig:299:28: 0xc88821 in std.Thread.MainFuncs.posixThreadMain (zig1)
                    startFn(arg);
                           ^
```

**Comment by zigazeljko** [comment]
Since #4335 is accepted, shouldn't this parse as a tuple type (with `A` being the field type instead of field name)?

**Comment by andrewrk** [comment]
Yes, however, self-hosted is being developed with mostly the same status quo as stage1, with the goal to switch to self-hosted as soon as possible. Once that is accomplished, language modifications will be implemented in both codebases at the same time to keep them in sync.

**Comment by andrewrk** [comment]
Landed in 31c49ad64ded02b9dde57f5d3ef102a771fa5cf7.
